In [1]:
## Current path display

import os
current_path = os.getcwd()
print(current_path)

d:\05_Projekte\04_Research\08_ManuBrain\01_Lab\Demo-Cell-Classification\02_Preprocessing


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime
from datetime import timedelta  

## Functions 

In [3]:
# return tag from data
def fc_tag (data,index):
    return data.at[index,'d'].split('=')[2] # [2] equivalent a tag column data en txt file

# update tag in main table
def fc_tag_update(data, tag):
    for i in range(len(tag)):
        i = i + 1 
        data = data.replace({'PrimaryKey':i}, fc_tag(tag,i), regex = False)
    return data

#Sync TimeStamp

def fc_sync_timestamp2(data,ServerTimeStamp,delta_time): #Requires df, column-name and delta-time
    for date in data[ServerTimeStamp]:
        source = datetime.fromisoformat(str(date).replace('Z','+00:00'))
        date_sync = source - delta_time
        data[ServerTimeStamp] = date_sync
    return data

def fc_sync_timestamp(data,ServerTimeStamp,delta_time): #Requires df, column-name and delta-time
    for i in range(len(data)): 
        source = datetime.fromisoformat(str(data.at[i,ServerTimeStamp]).replace('Z','+00:00'))
        date_sync = source - delta_time
        data.at[i,ServerTimeStamp] = date_sync
    return data

## Paths from data set

In [4]:
date = 201027
uc_num = 1

path = str(date) + '/use_case' + str(uc_num)

# Conveyor_belt
path_data_conv = '../01_Data/' + path + '/Conveyor Speed/DataLogger.csv'
path_tag_conv = '../01_Data/' + path + '/Conveyor Speed/tagconfig.txt'

# Gate
path_data_gate = '../01_Data/' + path + '/Gate/DataLogger.csv'
path_tag_gate = '../01_Data/' + path + '/Gate/tagconfig.txt'

# Quality_Control
path_data_qc = '../01_Data/' + path + '/Quality Control/DataLogger.csv'
path_tag_qc = '../01_Data/' + path + '/Quality Control/tagconfig.txt'

# Robot_Position
path_data_rp = '../01_Data/' + path + '/Robot Position/DataLogger.csv'
path_tag_rp = '../01_Data/' + path + '/Robot Position/tagconfig.txt'

# Store
path_data_st = '../01_Data/' + path + '/Store/DataLogger.csv'
path_tag_st = '../01_Data/' + path + '/Store/tagconfig.txt'

# Flag
path_data_flag = '../01_Data/' + path + '/Flag/DataLogger.csv'
path_tag_flag = '../01_Data/' + path + '/Flag/tagconfig.txt'

# Processed data

path_data_save = '../01_Data/' + path + '/Processed/'
path_sample_save = '../01_Data/' + path + '/Processed/Samples/'


## Import data from csv

In [5]:
# Conveyor_belt
data_conv =  pd.read_csv(path_data_conv, delimiter = ',')
#df_conv = pd.DataFrame(pd.read_csv(path_data_conv, delimiter = ','))

# Gate
data_gate =  pd.read_csv(path_data_gate, delimiter = ',')
#df_gate = pd.DataFrame(pd.read_csv(path_data_gate, delimiter = ','))

# Quality_Control
data_qc =  pd.read_csv(path_data_qc, delimiter = ',')
#df_qc = pd.DataFrame(pd.read_csv(path_data_qc, delimiter = ','))

# Robot_Position
data_rp =  pd.read_csv(path_data_rp, delimiter = ',')
#df_rp = pd.DataFrame(pd.read_csv(path_data_rp, delimiter = ','))

# Store
data_st =  pd.read_csv(path_data_st, delimiter = ',')
#df_st = pd.DataFrame(pd.read_csv(path_data_st, delimiter = ','))

# Flag
data_flag =  pd.read_csv(path_data_flag, delimiter = ',')
#df_flag = pd.DataFrame(pd.read_csv(path_data_st, delimiter = ','))

## Import data tags

In [6]:
# Conveyor_belt
tag_conv =  pd.read_fwf(path_tag_conv, sep=" ", header=None, names=['a','b','c','d'])
tag_conv = tag_conv.drop([0]) # delete first row (trash)

# Gate
tag_gate =  pd.read_fwf(path_tag_gate, sep=" ", header=None, names=['a','b','c','d'])
tag_gate = tag_gate.drop([0]) # delete first row (trash)

# Quality_Control
tag_qc =  pd.read_fwf(path_tag_qc, sep=" ", header=None, names=['a','b','c','d'])
tag_qc = tag_qc.drop([0]) # delete first row (trash)

# Robot_Position
tag_rp =  pd.read_fwf(path_tag_rp, sep=" ", header=None, names=['a','b','c','d'])
tag_rp = tag_rp.drop([0]) # delete first row (trash)

# Robot_Position
tag_st =  pd.read_fwf(path_tag_st, sep=" ", header=None, names=['a','b','c','d'])
tag_st = tag_st.drop([0]) # delete first row (trash)

# Flag
tag_flag =  pd.read_fwf(path_tag_flag, sep=" ", header=None, names=['a','b','c','d'])
tag_flag = tag_flag.drop([0]) # delete first row (trash)

## Update tags

In [7]:
# Conveyor_belt
data_conv = fc_tag_update(data_conv, tag_conv)

# Gate
data_gate = fc_tag_update(data_gate, tag_gate)

# Quality_Control
data_qc = fc_tag_update(data_qc, tag_qc)

# Robot_Position
data_rp = fc_tag_update(data_rp, tag_rp)

# Store
data_st = fc_tag_update(data_st, tag_st)

# Flag
data_flag = fc_tag_update(data_flag, tag_flag)

## Syncronize ServerTimeStamp

In [8]:
delta_time_PLC = timedelta(seconds=67)
delta_time_SIM = timedelta(seconds=0)

In [9]:
data_conv = fc_sync_timestamp(data_conv,'ServerTimeStamp',delta_time_PLC)
data_conv

Empty DataFrame
Columns: [PrimaryKey, DataType, Value, StatusCode, SourceTimeStamp, ServerTimeStamp]
Index: []

In [10]:
data_gate = fc_sync_timestamp(data_gate,'ServerTimeStamp',delta_time_SIM)
data_gate

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0          Speed   0x000b    -30  0x00000000  2020-10-27T15:05:24.596Z   
1          Speed   0x000b    -30  0x00000000  2020-10-27T15:05:24.896Z   
2           Safe   0x0001  false  0x00000000  2020-10-27T15:05:24.596Z   
3           Safe   0x0001  false  0x00000000  2020-10-27T15:05:24.896Z   
4       Position   0x000b   31.6  0x00000000  2020-10-27T15:05:24.596Z   
...          ...      ...    ...         ...                       ...   
21517   Position   0x000b    0.3  0x00000000  2020-10-27T15:41:15.896Z   
21518   Position   0x000b    0.3  0x00000000  2020-10-27T15:41:16.196Z   
21519      Speed   0x000b      0  0x00000000  2020-10-27T15:41:16.496Z   
21520       Safe   0x0001   true  0x00000000  2020-10-27T15:41:16.496Z   
21521   Position   0x000b    0.3  0x00000000  2020-10-27T15:41:16.496Z   

                        ServerTimeStamp  
0      2020-10-27 15:05:24.596000+00:00  
1      2020-10-27 15:05:24.896000+00:00  
2      2020-10-27 15:05:24.596000+00:00  
3      2020-10-27 15:05:24.896000+00:00  
4      2020-10-27 15:05:24.596000+00:00  
...                                 ...  
21517  2020-10-27 15:41:15.896000+00:00  
21518  2020-10-27 15:41:16.196000+00:00  
21519  2020-10-27 15:41:16.496000+00:00  
21520  2020-10-27 15:41:16.496000+00:00  
21521  2020-10-27 15:41:16.496000+00:00  

[21522 rows x 6 columns]

In [11]:
data_qc = fc_sync_timestamp(data_qc,'ServerTimeStamp',delta_time_SIM)
data_qc

PrimaryKey DataType     Value  StatusCode           SourceTimeStamp  \
0      DistanceAbs   0x000b  0.000000  0x00000000  2020-10-27T15:05:26.732Z   
1      DistanceAbs   0x000b  0.000000  0x00000000  2020-10-27T15:05:27.032Z   
2        DistanceX   0x000b  0.000000  0x00000000  2020-10-27T15:05:26.732Z   
3        DistanceX   0x000b  0.000000  0x00000000  2020-10-27T15:05:27.032Z   
4        DistanceY   0x000b  0.000000  0x00000000  2020-10-27T15:05:26.732Z   
...            ...      ...       ...         ...                       ...   
28687    DistanceX   0x000b  0.433815  0x00000000  2020-10-27T15:41:18.332Z   
28688    DistanceY   0x000b  0.215054  0x00000000  2020-10-27T15:41:18.032Z   
28689    DistanceY   0x000b  0.215054  0x00000000  2020-10-27T15:41:18.332Z   
28690   ResultCode   0x0006  0.000000  0x00000000  2020-10-27T15:41:18.032Z   
28691   ResultCode   0x0006  0.000000  0x00000000  2020-10-27T15:41:18.332Z   

                        ServerTimeStamp  
0      2020-10-27 15:05:26.732000+00:00  
1      2020-10-27 15:05:27.032000+00:00  
2      2020-10-27 15:05:26.732000+00:00  
3      2020-10-27 15:05:27.032000+00:00  
4      2020-10-27 15:05:26.732000+00:00  
...                                 ...  
28687  2020-10-27 15:41:18.332000+00:00  
28688  2020-10-27 15:41:18.032000+00:00  
28689  2020-10-27 15:41:18.332000+00:00  
28690  2020-10-27 15:41:18.032000+00:00  
28691  2020-10-27 15:41:18.332000+00:00  

[28692 rows x 6 columns]

In [12]:
data_rp = fc_sync_timestamp(data_rp,'ServerTimeStamp',delta_time_PLC)
data_rp

PrimaryKey DataType       Value  StatusCode           SourceTimeStamp  \
0            527   0x0007  1050956941  0x00000000  2020-10-27T15:06:35.382Z   
1            527   0x0007  1050958021  0x00000000  2020-10-27T15:06:35.628Z   
2            528   0x0007  3190244164  0x00000000  2020-10-27T15:06:35.382Z   
3            528   0x0007  3190243986  0x00000000  2020-10-27T15:06:35.628Z   
4            529   0x0007  1051537273  0x00000000  2020-10-27T15:06:35.385Z   
...          ...      ...         ...         ...                       ...   
25665        527   0x0007  3196290032  0x00000000  2020-10-27T15:42:26.627Z   
25666        528   0x0007  3188549527  0x00000000  2020-10-27T15:42:26.377Z   
25667        528   0x0007  3190340140  0x00000000  2020-10-27T15:42:26.625Z   
25668        529   0x0007  1054308511  0x00000000  2020-10-27T15:42:26.374Z   
25669        529   0x0007  1054601533  0x00000000  2020-10-27T15:42:26.624Z   

                        ServerTimeStamp  
0      2020-10-27 15:05:28.382000+00:00  
1      2020-10-27 15:05:28.628000+00:00  
2      2020-10-27 15:05:28.382000+00:00  
3      2020-10-27 15:05:28.628000+00:00  
4      2020-10-27 15:05:28.385000+00:00  
...                                 ...  
25665  2020-10-27 15:41:19.627000+00:00  
25666  2020-10-27 15:41:19.377000+00:00  
25667  2020-10-27 15:41:19.625000+00:00  
25668  2020-10-27 15:41:19.374000+00:00  
25669  2020-10-27 15:41:19.624000+00:00  

[25670 rows x 6 columns]

In [13]:
data_st = fc_sync_timestamp(data_st,'ServerTimeStamp',delta_time_SIM)
data_st

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0      DiscSize   0x0003      2  0x00000000  2020-10-27T15:05:30.485Z   
1      DiscSize   0x0003      2  0x00000000  2020-10-27T15:05:30.785Z   
2      DiscSize   0x0003      2  0x00000000  2020-10-27T15:05:31.085Z   
3      DiscSize   0x0003      2  0x00000000  2020-10-27T15:05:31.385Z   
4      DiscSize   0x0003      2  0x00000000  2020-10-27T15:05:31.685Z   
...         ...      ...    ...         ...                       ...   
7168   DiscSize   0x0003      1  0x00000000  2020-10-27T15:41:20.885Z   
7169   DiscSize   0x0003      1  0x00000000  2020-10-27T15:41:21.185Z   
7170   DiscSize   0x0003      1  0x00000000  2020-10-27T15:41:21.485Z   
7171   DiscSize   0x0003      1  0x00000000  2020-10-27T15:41:21.785Z   
7172   DiscSize   0x0003      1  0x00000000  2020-10-27T15:41:22.085Z   

                       ServerTimeStamp  
0     2020-10-27 15:05:30.485000+00:00  
1     2020-10-27 15:05:30.785000+00:00  
2     2020-10-27 15:05:31.085000+00:00  
3     2020-10-27 15:05:31.385000+00:00  
4     2020-10-27 15:05:31.685000+00:00  
...                                ...  
7168  2020-10-27 15:41:20.885000+00:00  
7169  2020-10-27 15:41:21.185000+00:00  
7170  2020-10-27 15:41:21.485000+00:00  
7171  2020-10-27 15:41:21.785000+00:00  
7172  2020-10-27 15:41:22.085000+00:00  

[7173 rows x 6 columns]

In [14]:
data_flag = fc_sync_timestamp(data_flag,'ServerTimeStamp',delta_time_PLC)
data_flag

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0          561   0x0001   true  0x00000000  2020-10-27T15:06:29.626Z   
1          560   0x0004      0  0x00000000  2020-10-27T15:06:29.629Z   
2          560   0x0004      1  0x00000000  2020-10-27T15:06:35.128Z   
3          561   0x0001  false  0x00000000  2020-10-27T15:06:35.878Z   
4          560   0x0004      2  0x00000000  2020-10-27T15:06:41.627Z   
..         ...      ...    ...         ...                       ...   
623        560   0x0004      3  0x00000000  2020-10-27T15:42:03.378Z   
624        560   0x0004      0  0x00000000  2020-10-27T15:42:10.136Z   
625        561   0x0001   true  0x00000000  2020-10-27T15:42:10.879Z   
626        560   0x0004      1  0x00000000  2020-10-27T15:42:17.379Z   
627        561   0x0001  false  0x00000000  2020-10-27T15:42:18.126Z   

                      ServerTimeStamp  
0    2020-10-27 15:05:22.626000+00:00  
1    2020-10-27 15:05:22.629000+00:00  
2    2020-10-27 15:05:28.128000+00:00  
3    2020-10-27 15:05:28.878000+00:00  
4    2020-10-27 15:05:34.627000+00:00  
..                                ...  
623  2020-10-27 15:40:56.378000+00:00  
624  2020-10-27 15:41:03.136000+00:00  
625  2020-10-27 15:41:03.879000+00:00  
626  2020-10-27 15:41:10.379000+00:00  
627  2020-10-27 15:41:11.126000+00:00  

[628 rows x 6 columns]

## Merge Data Sets

In [15]:
data_set = [data_conv, data_gate, data_qc, data_rp, data_st, data_flag]
data_merge = pd.concat(data_set, keys=['conv', 'gate', 'qc', 'rp', 'store', 'flag'])
# Sort the data_set by SourceTimeStamp, reset index and delete aux column
data_merge = data_merge.sort_values(by='ServerTimeStamp').reset_index().drop('level_1',1)
#del data_merge['level_1'] #delete aux column

In [16]:
data_merge

level_0 PrimaryKey DataType  Value  StatusCode  \
0        flag        561   0x0001   true  0x00000000   
1        flag        560   0x0004      0  0x00000000   
2        gate      Speed   0x000b    -30  0x00000000   
3        gate       Safe   0x0001  false  0x00000000   
4        gate   Position   0x000b   31.6  0x00000000   
...       ...        ...      ...    ...         ...   
83680   store   DiscSize   0x0003      1  0x00000000   
83681   store   DiscSize   0x0003      1  0x00000000   
83682   store   DiscSize   0x0003      1  0x00000000   
83683   store   DiscSize   0x0003      1  0x00000000   
83684   store   DiscSize   0x0003      1  0x00000000   

                SourceTimeStamp                   ServerTimeStamp  
0      2020-10-27T15:06:29.626Z  2020-10-27 15:05:22.626000+00:00  
1      2020-10-27T15:06:29.629Z  2020-10-27 15:05:22.629000+00:00  
2      2020-10-27T15:05:24.596Z  2020-10-27 15:05:24.596000+00:00  
3      2020-10-27T15:05:24.596Z  2020-10-27 15:05:24.596000+00:00  
4      2020-10-27T15:05:24.596Z  2020-10-27 15:05:24.596000+00:00  
...                         ...                               ...  
83680  2020-10-27T15:41:20.885Z  2020-10-27 15:41:20.885000+00:00  
83681  2020-10-27T15:41:21.185Z  2020-10-27 15:41:21.185000+00:00  
83682  2020-10-27T15:41:21.485Z  2020-10-27 15:41:21.485000+00:00  
83683  2020-10-27T15:41:21.785Z  2020-10-27 15:41:21.785000+00:00  
83684  2020-10-27T15:41:22.085Z  2020-10-27 15:41:22.085000+00:00  

[83685 rows x 7 columns]

## Export Data to csv

In [17]:
data_merge.to_csv(path_data_save + 'data_merge.csv')
print('The data is saved at ' + path_data_save + 'data_merge.csv')

The data is saved at ../01_Data/201027/use_case1/Processed/data_merge.csv


In [18]:
len(data_merge)

83685

## Determine the starting point of data set

#Based on change of DistanceAbs in Data- Set
var_actual = 0
run_list = [0]
for i in range(len(data_merge)):
    var_previous = var_actual
    if data_merge.at[i,'PrimaryKey'] == 'DistanceAbs':  # if 'PrimaryKey' value = 'DistanceAbs':
        var_actual = data_merge.at[i,'Value'] # value of current DistanceAbs
    if var_previous != var_actual:
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample    

# Based on grabbing the white disk
var_actual = 99
disk_actual = 99 # Initial value
run_list = [0]
disk = False
storage_down = False
rob_x = 0
rob_y = 0
rob_z = 0
rob_pos = 0
pos = [0]*(len(data_merge)-2)
for i in range(len(data_merge)-2): # len(data_merge)-2 is considered to avoid border errors
    #var_previous = var_actual
    #rob_pos = fc_rob_pos(i, data_merge) # review robot position
    pos[i] = 0
    if data_merge.at[i,'PrimaryKey'] == 'DiscSize':  # if 'PrimaryKey' value = 'DiskSize':
        disk_previous = disk_actual
        disk_actual = data_merge.at[i,'Value'] # value of current DiskSize
        #print(disk_previous, disk_actual, i)
        if disk_previous != 2 and disk_actual == 2: # Varialbe 0: unkown; 1: Black disk; 2: white disk
            disk = True
            #print('New white disk here ' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '527':  # if 'PrimaryKey' value = 'x coordinate':
        rob_x = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('x = ' + str(rob_x) + '--' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '528':  # if 'PrimaryKey' value = 'y coordinate':
        rob_y = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('y = ' + str(rob_y) + '--' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '529':  # if 'PrimaryKey' value = 'z coordinate':
        rob_z = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('z = ' + str(rob_z) + '--' + str(i))
    if rob_x > 3100000000 and rob_y > 3100000000 and rob_z < 1051200000: #Robot in storage. Picking up the disk
        rob_pos = 1
        pos[i] = 1
        #print('pos 1 - ' + str(i))
    if rob_x < 1052000000 and rob_y > 3100000000 and rob_z < 1052000000 and rob_z > 1051200000 : #Robot in storage. Placing down the disk
        rob_pos = 2
        pos[i] = 2
        #print('pos 2 - ' + str(i))
    #print(rob_pos)
    if rob_pos == 1: # Robot picking up a disk
        storage_down = True
        #print(storage_down,i)  
    if disk == True and storage_down == True:
        #var_actual = rob_pos
        disk = False
        storage_down = False
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample
        #print(i)
    #if var_previous != var_actual:
        #run_list.append(i) #i append to list # This list will tell at how many i contains one sample   
 

In [19]:
# Based on grabbing the white disk
run_list = [0]
rob_pos_actual = 99
rob_pos_old = 0
gripper_actual = 99
gripper_old = 0
gripper = False
rob_pos = False
for i in range(len(data_merge)-2): # len(data_merge)-2 is considered to avoid border errors
    #var_previous = var_actual
    #rob_pos = fc_rob_pos(i, data_merge) # review robot position
    if data_merge.at[i,'PrimaryKey'] == '561':  # if 'PrimaryKey' value = 'gripper open (0), close (1)':
        gripper_old = gripper_actual
        gripper_actual = data_merge.at[i,'Value'] # value of current 'x coordinate'
        if gripper_old != gripper_actual:
            gripper = True
    if data_merge.at[i,'PrimaryKey'] == '560':  # if 'PrimaryKey' value = 'position of robot based in certain conditions':
        # 0: Black disk in store
        # 1: Black disk in gate
        # 2: White disk in store
        # 3: White disk in gate
        rob_pos_old = rob_pos_actual
        rob_pos_actual = int(data_merge.at[i,'Value']) # value of current 'x coordinate'
        if rob_pos_old != rob_pos_actual and rob_pos_actual == 2:
            rob_pos = True
    if rob_pos == True and gripper == True:
        rob_pos = False
        gripper = False
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample
        #print(i)
 

In [20]:
print(run_list) # Starting points. 
print(len(run_list))

[0, 303, 1355, 2418, 3470, 4529, 5581, 6641, 7693, 8756, 9808, 10870, 11922, 12985, 14876, 15928, 16980, 18043, 19095, 20154, 21206, 22266, 23318, 24381, 25433, 26495, 27558, 28609, 29672, 30725, 31783, 32854, 33906, 34968, 36020, 37094, 38146, 39213, 40265, 41328, 42380, 43442, 44494, 45557, 46609, 47672, 48724, 49784, 50835, 51895, 52947, 54010, 55062, 56124, 57187, 58240, 59302, 60354, 61406, 62469, 63525, 64577, 65641, 66692, 67766, 68817, 69880, 70932, 71991, 73043, 74103, 75155, 76218, 77271, 78333, 79385, 80436, 81499, 82551]
79


In [21]:
# data_merge.iloc[list(range(0,3))]

In [23]:
#df_robpos = pd.DataFrame(pos,columns=['RobPos']) # Position of robot. 0: not important; 1: Robot in storage. Picking up the disk; 2:Robot in storage. Placing down the disk
#data_merge = pd.concat([data_merge, df_robpos], axis=1, sort=False)
data_merge

level_0 PrimaryKey DataType  Value  StatusCode  \
0        flag        561   0x0001   true  0x00000000   
1        flag        560   0x0004      0  0x00000000   
2        gate      Speed   0x000b    -30  0x00000000   
3        gate       Safe   0x0001  false  0x00000000   
4        gate   Position   0x000b   31.6  0x00000000   
...       ...        ...      ...    ...         ...   
83680   store   DiscSize   0x0003      1  0x00000000   
83681   store   DiscSize   0x0003      1  0x00000000   
83682   store   DiscSize   0x0003      1  0x00000000   
83683   store   DiscSize   0x0003      1  0x00000000   
83684   store   DiscSize   0x0003      1  0x00000000   

                SourceTimeStamp                   ServerTimeStamp  
0      2020-10-27T15:06:29.626Z  2020-10-27 15:05:22.626000+00:00  
1      2020-10-27T15:06:29.629Z  2020-10-27 15:05:22.629000+00:00  
2      2020-10-27T15:05:24.596Z  2020-10-27 15:05:24.596000+00:00  
3      2020-10-27T15:05:24.596Z  2020-10-27 15:05:24.596000+00:00  
4      2020-10-27T15:05:24.596Z  2020-10-27 15:05:24.596000+00:00  
...                         ...                               ...  
83680  2020-10-27T15:41:20.885Z  2020-10-27 15:41:20.885000+00:00  
83681  2020-10-27T15:41:21.185Z  2020-10-27 15:41:21.185000+00:00  
83682  2020-10-27T15:41:21.485Z  2020-10-27 15:41:21.485000+00:00  
83683  2020-10-27T15:41:21.785Z  2020-10-27 15:41:21.785000+00:00  
83684  2020-10-27T15:41:22.085Z  2020-10-27 15:41:22.085000+00:00  

[83685 rows x 7 columns]

The list contains the point i when the value is different. To consider the point where the value remains, consider i-1

In [24]:
d = {} # Dictionary that holds the sub-data-sets given by the starting point
for i in range(len(run_list)-1): # Ignores the last value (set of values)
    list_range = list(range(run_list[i],run_list[i+1]))
    d['df{0}'.format(i)] = data_merge.iloc[list_range]
    
print('Dictionary lenght is:' + str(len(d)))

Dictionary lenght is:78


## Assure conv_speed per sample 

In [25]:
conv_1 = 0
conv_2 = 0
conv_3 = 0
conv1 = False
conv2 = False
conv3 = False
for j in range(len(d)): # as many samples saved in dictionary
    df = d['df'+ str(j)]
    for i in range(len(df)): # as many row in the dataFrame
        #print(df.iat[i,1])
        if df.iat[i,1] == '27': # conv1
            conv_1 = df.iat[i,3]
            conv1 = True # the value shows at least once in the data set
        if df.iat[i,1] == '28': # conv2
            conv_2 = df.iat[i,3]
            conv2 = True
        if df.iat[i,1] == '29': # conv3
            conv_3 = df.iat[i,3]
            conv3 = True
    if conv1 == False: # The value did not occur in this sample. Copy value from previous samples into this one
        dic = {'PrimaryKey' : '27', 'Value': str(conv_1)}
        df.loc[len(df)+1,'PrimaryKey'] = '27'
        df.loc[len(df),'Value'] = str(conv_1)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
    if conv2 == False:
        dic = {'PrimaryKey' : '27', 'Value': str(conv_2)}
        df.loc[len(df)+1,'PrimaryKey'] = '28'
        df.loc[len(df),'Value'] = str(conv_2)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
    if conv3 == False:
        dic = {'PrimaryKey' : '27', 'Value': str(conv_3)}
        df.loc[len(df)+1,'PrimaryKey'] = '29'
        df.loc[len(df),'Value'] = str(conv_3)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
   
    conv1 = False # reset conv for the next round
    conv2 = False
    conv3 = False
print(conv_1)
print(conv_2)
print(conv_3)

0
0
0


## Export individual data-sample to .csv

In [26]:
for i in range(len(d)):
    if i > 0: # always ignore the first sample due to data-acquisition misalligment
        d['df' + str(i)].to_csv(path_sample_save + 'sample_' + str(i) + '.csv')
print('The data is saved at ' + path_sample_save + 'sample_x.csv')

The data is saved at ../01_Data/201027/use_case1/Processed/Samples/sample_x.csv


## Testing area

In [27]:
#d['df67']

In [28]:
# list_range = list(range(run_list[0],run_list[1]))
# print(list_range)

In [29]:
# data_merge.iloc[list_range]

In [30]:
# data_rp.loc[:, 'PrimaryKey'] = '529_z'
# data_rp

In [31]:
# %matplotlib notebook 
# data_gate.head(400).plot(x='SourceTimeStamp', y= 'Value')
# plt.xticks(rotation=45)

In [32]:
# %matplotlib notebook 
# data_rp.head(200).plot( y= 'Value')
# plt.xticks(rotation=45)